### PR TITLE
xds: refactor XdsLoadReportClient and XdsLoadStatsStore in order to integrate with XdsLoadBalancer (part 1)

### PR DESCRIPTION
**Pull request by voidzcy**
This PR performs some pre-work for integrating client side load reporting with xds load balancer. Originally, this is part of #5836. For better readability and easier to review, #5836 is split into two parts:
1. Make necessary changes and refactoring in load reporting part to be compatible with xds load balancer implementation. (This PR)
2. Integrate load reporting with xds load balancer, with tests verifying load reporting APIs are invoked by xds load balancer properly. (Next PR)

Main changes in this PR:
- Put self-defined `Locality` class into a separate file and rename it to `XdsLocality`. It is used both in load reporting and xds load balancer.
- Deleted delegation calls of `addLocality()`, `removeLocality()`, `interceptPickResult()` and `recordDroppedRequest` in `XdsLoadReportClient`. EDS response handler in xds load balancer will directly interact with `StatsStore` for this methods, so no longer invoke them through `XdsLoadReportClient`.
- ~Fixed bug of using the wrong `cluster_name` in `ClusterStats` of `LoadStatsRequest`. This should be the GSLB service name, instead of the load-balanced service name (i.e., hostname) from client perspective. This `cluster_name` is received in `LoadStatsResponse` in load reporting response (as well as in EDS response). See [grpc-go](https://github.com/grpc/grpc-go/pull/2845)'s change as well.~

### CODE REVIEW

File: xds/src/main/java/io/grpc/xds/XdsLocality.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+  @Override
+  public String toString() {
+    return MoreObjects.toStringHelper(this)
+        .add("region", region)
+        .add("zone", zone)
+        .add("subzone", subzone)
+        .toString();
+  }
+}
```
Comment (dapengzhang0) [inline xds/src/main/java/io/grpc/xds/XdsLocality.java]: nit: add a new line

File: xds/src/main/java/io/grpc/xds/StatsStore.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+  StatsCounter getLocalityCounter(XdsLocality locality);
+
+  /**
+   * Records a drop decision made by a {@link io.grpc.LoadBalancer.SubchannelPicker} instance
+   * with the provided category. Drops are aggregated in service level.
+   *
+   * <p>This method is thread-safe.
+   */
+  void recordDroppedRequest(String category);
+}
```
Comment (dapengzhang0) [inline xds/src/main/java/io/grpc/xds/StatsStore.java]: nit: add new line

File: xds/src/main/java/io/grpc/xds/XdsLocality.java (hunk 1 above)
Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/XdsLocality.java]: Thanks. Done.

File: xds/src/main/java/io/grpc/xds/StatsStore.java (hunk 2 above)
Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/StatsStore.java]: Done.

**Comment by voidzcy** [comment]
> Can we split the refactor and bug fix? Let's merge refactor first.

Sure.

File: xds/src/test/java/io/grpc/xds/XdsLocalityTest.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    XdsLocality xdsLocality = XdsLocality.fromLocalityProto(locality);
+    assertThat(xdsLocality.getRegion()).isEqualTo("test_region");
+    assertThat(xdsLocality.getZone()).isEqualTo("test_zone");
+    assertThat(xdsLocality.getSubzone()).isEqualTo("test_subzone");
+
+    Locality convertedLocality = xdsLocality.toLocalityProto();
+    assertThat(convertedLocality.getRegion()).isEqualTo("test_region");
+    assertThat(convertedLocality.getZone()).isEqualTo("test_zone");
+    assertThat(convertedLocality.getSubZone()).isEqualTo("test_subzone");
+  }
```
Comment (zhangkun83) [inline xds/src/test/java/io/grpc/xds/XdsLocalityTest.java]: Please test `equals()` like in:
https://github.com/grpc/grpc-java/blob/9895e243b1651d03a561e75ec7dd4fd8f16cc49f/core/src/test/java/io/grpc/inprocess/InProcessSocketAddressTest.java#L30

Also test `hashCode()`.

File: xds/src/main/java/io/grpc/xds/StatsStore.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+  /**
+   * Returns the {@link StatsCounter} that does locality level stats aggregation for the provided
+   * locality. It is caller's responsibility to ensure the counter exists
+   * (i.e., after {@code #addLocality} and before {@code #removeLocality} is called for the
+   * provided locality). Otherwise, {@code null} will be returned.
+   *
+   * <p>This method is thread-safe.
+   */
+  StatsCounter getLocalityCounter(XdsLocality locality);
```
[the remark this answers is not in the document]
Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/StatsStore.java]: Done.

File: xds/src/main/java/io/grpc/xds/StatsStore.java [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+   *
+   * <p>This method is not thread-safe and should be called from the same synchronized context *
+   * returned by {@link XdsLoadBalancer.Helper#getSynchronizationContext}.
+   */
+  void removeLocality(XdsLocality locality);
+
+  /**
+   * Applies client side load recording to {@link PickResult}s picked by the intra-locality picker
+   * for the provided locality.
+   *
```
[the remark this answers is not in the document]
Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/StatsStore.java]: Done.

File: xds/src/test/java/io/grpc/xds/XdsLocalityTest.java (hunk 3 above)
Comment (voidzcy) [reply xds/src/test/java/io/grpc/xds/XdsLocalityTest.java]: Done.

File: xds/src/main/java/io/grpc/xds/StatsStore.java [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+   * <p>This method should be called in the same synchronized context that
+   * {@link XdsLoadBalancer.Helper#getSynchronizationContext} returns.
+   */
+  ClusterStats generateLoadReport();
+
+  /**
+   * Tracks load stats for endpoints in the provided locality. To be called upon balancer locality
+   * updates only for newly assigned localities. Only load stats for endpoints in added localities
+   * will be reported to the remote balancer. This method needs to be called at locality updates
+   * only for * newly assigned localities in balancer discovery responses.
```
[the remark this answers is not in the document]
Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/StatsStore.java]: Done

File: xds/src/main/java/io/grpc/xds/StatsStore.java [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+   *
+   * <p>This method is thread-safe.
+   */
+  PickResult interceptPickResult(PickResult pickResult, XdsLocality locality);
+
+  /**
+   * Returns the {@link StatsCounter} that does locality level stats aggregation for the provided
+   * locality. It is caller's responsibility to ensure the counter exists
+   * (i.e., after {@code #addLocality} and before {@code #removeLocality} is called for the
+   * provided locality). Otherwise, {@code null} will be returned.
```
[the remark this answers is not in the document]
Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/StatsStore.java]: Done.

**Comment by dapengzhang0** [comment]
@voidzcy Thanks for the refactoring.

**Pull request merged**